### PR TITLE
Utilize ComPtr in D3D, DInput and XAudio2 backends

### PR DESCRIPTION
**Pull request by beholdnec**
This PR uses COM smart pointers to manage COM pointers.

This makes it easier to avoid resource leaks of COM objects.

### CODE REVIEW

File: Source/Core/InputCommon/ControllerInterface/DInput/DInputJoystick.cpp [hunk 1]
```diff
@@ -46,28 +46,25 @@ void InitJoystick(IDirectInput8* const idi8, HWND hwnd)
                   js_device->SetCooperativeLevel(nullptr, DISCL_BACKGROUND | DISCL_NONEXCLUSIVE)))
           {
             // PanicAlert("SetCooperativeLevel failed!");
-            js_device->Release();
             continue;
           }
         }
 
-        auto js = std::make_shared<Joystick>(js_device);
+        auto js = std::make_shared<Joystick>(js_device.Get());
```
Comment (iwubcode) [inline Source/Core/InputCommon/ControllerInterface/DInput/DInputJoystick.cpp]: Like I said in the other PR, this should probably be passed all the way in if the device is just going to take the raw pointer and then create a shared pointer from it

File: Source/Core/VideoBackends/D3D/D3DShader.cpp [hunk 2]
```diff
@@ -177,51 +174,43 @@ bool CompilePixelShader(const std::string& code, D3DBlob** blob, const D3D_SHADE
                D3D::PixelShaderVersionString(), (const char*)errorBuffer->GetBufferPointer());
 
     *blob = nullptr;
-    errorBuffer->Release();
   }
   else
   {
-    *blob = new D3DBlob(shaderBuffer);
-    shaderBuffer->Release();
+    *blob = new D3DBlob(shaderBuffer.Get());
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/D3DShader.cpp]: Sorry for being repetitive but not sure why we are bothering to pass in a raw pointer, we're upping the ref count on the new smart ptr and deallocating the smart ptr in this scope...

File: Source/Core/InputCommon/ControllerInterface/DInput/DInputJoystick.cpp (hunk 1 above)
Comment (beholdnec) [inline Source/Core/InputCommon/ControllerInterface/DInput/DInputJoystick.cpp]: Whoops! I meant to change that line. Good catch.

**Comment by beholdnec** [comment]
Updated with minor changes.

The rule I've been using for smart-pointer vs. raw-pointer is: If a function is supposed to take ownership of the object or if the argument could be stored somewhere, pass an object by a ComPtr. Otherwise, if the function only references an object without storing or taking ownership, pass by raw pointer.

**Comment by iwubcode** [comment]
Fair enough.  I still think there's some benefit to passing by const ref most of the time, since then the user doesn't have to worry about the lifetime of the object and there's really a very minor penalty.

However, I'd be fine with the code as is, it's at least an improvement over the old stuff.  LGTM from me.

**Comment by beholdnec** [comment]
Update, hopefully for the final time: Pass COM objects by smart-pointer in more instances.

Changing some function arguments to "moved" ComPtr's, e.g. InitializeFromExisting(ComPtr<ID3D11Texture2D>&& tex), allows the compiler to remove an AddRef/Release cycle entirely.

**Comment by beholdnec** [comment]
Update: Fix styling issue, rebase on master.

**Comment by beholdnec** [comment]
Update: Rebase on master, fix style issue.

**Comment by beholdnec** [comment]
Update: Rebase on master, fix style issues.

I've been running into lint problems because I was using clang-format 4.0.1, which has slightly different rules from clang-format 3.9.1. I'm crossing my fingers that this works.

**Comment by leoetlino** [comment]
For the information, the version we currently use is 3.8.

**Comment by beholdnec** [comment]
Update: Rebased on master.

**Comment by beholdnec** [comment]
Please do not merge, there is some unexpected breakage in EFB->RAM and XFB encoders.

**Comment by beholdnec** [comment]
Update: Fix EFB->RAM and XFB encoders. This broke because taking a pointer to a ComPtr causes the pointer to be released and set to null. That's not how I would have designed ComPtr, but we work with what we've got.

**Comment by magcius** [comment]
Can't you just invent a better COM pointer that does the Release for you?

**Comment by beholdnec** [comment]
I actually did, but we opted to use the one provided by Microsoft!

**Comment by stenzek** [comment]
I think PR #7869 covered most of the changes here. Please feel free to re-open if I missed any instances.

**Pull request closed without merging**